rust/automerge/src/automerge.rs

```diff
@@ -805,23 +805,27 @@ impl Automerge {
                     ));
                 }
                 if self.is_causally_ready(&c) {
-                    self.apply_change(c, &mut op_observer);
+                    self.apply_change(c, &mut op_observer)?;
                 } else {
                     self.queue.push(c);
                 }
             }
         }
         while let Some(c) = self.pop_next_causally_ready_change() {
             if !self.history_index.contains_key(&c.hash()) {
-                self.apply_change(c, &mut op_observer);
+                self.apply_change(c, &mut op_observer)?;
             }
         }
         Ok(())
     }
 
-    fn apply_change<Obs: OpObserver>(&mut self, change: Change, observer: &mut Option<&mut Obs>) {
+    fn apply_change<Obs: OpObserver>(
+        &mut self,
+        change: Change,
+        observer: &mut Option<&mut Obs>,
+    ) -> Result<(), AutomergeError> {
```

Comment:
I don't think this function should return a `Result`. `apply_change` is a private function which is only ever intended to be called with a change which is "causally ready". You can see this in e.g. `apply_change_with`, where we use `is_causally_ready` or `pop_next_causally_ready` to make sure that the change dependencies are available before applying it and if the dependencies are not ready we put the change into `self.queue`.

As you've discovered, these checks were _not_ being made in `load`. I think the solution would be to add these checks there.

Reply:
:+1: I'll remove this fix from this branch to unblock the rest, and resubmit in a few days